### ISSUE TITLE
MapStruct doesn't respect interfaces with record-style getters

### ISSUE THREAD

**Issue by Akatroj** [mapstruct/mapstruct#3649]
### Expected behavior

Hello, 
I tried using a mapper that accepts an interface with record-like getters (`field()` instead of `getField()`) as an argument and MapStruct did not generate the correct code (left the fields uninitialized instead of using the getter). I provided code example below.

### Steps to reproduce the problem

mapper code:
```java
@Mapper
public interface TestMapper {
    TestMapper INSTANCE = Mappers.getMapper(TestMapper.class);

    SomeModelWithUserId fromUserIdAndSomeOtherData(UserIdProvider userIdProvider, String someOtherData);
}
```

target model:
```java
public record SomeModelWithUserId(
    long userId,
    String someOtherData
) {
}
```

`UserIdProvider` Interface:
```java
public interface UserIdProvider {
    long userId();
}
```

record implementing `UserIdProvider` :
```java
public record User(
    long userId
) implements UserIdProvider {
}
```

example usage of the mapper

```java
System.out.println(
    TestMapper.INSTANCE.fromUserIdAndSomeOtherData(
        new User(1),
        "someOtherData"
    )
);
```

The generated code should use `UserIdProvider::userId` method to set the userId, but it does not. Instead, it leaves `userId` uninitialized (`long userId = 0L;`) and gives this warning during build.

```
TestMapper.java:10: warning: Unmapped target property: "userId".
    SomeModelWithUserId fromUserIdAndSomeOtherData(UserIdProvider userIdProvider, String someOtherData);
```

I believe this is a bug, because:

 * If the `UserIdProvider` interface had a `getUserId()` method instead of `userId()`, it would work:

```java
public interface UserIdProvider {
    long userId();

    // with this, mapstruct generates correct code
    default long getUserId() {
        return userId();
    }
}
```

* if the type of the argument in the mapper was a record implementing `UserIdProvider`, and not `UserIdProvider` interface itself, it would work (using the `userId()` record getter):
```java
@Mapper
public interface TestMapper {
    TestMapper INSTANCE = Mappers.getMapper(TestMapper.class);

    //works
    SomeModelWithUserId fromUserIdAndSomeOtherData(User user, String someOtherData);
}
```

### MapStruct Version

1.5.5.Final and 1.6.0RC1

**Comment by zyberzebra** [comment]
I don't see something in the Documentation https://mapstruct.org/documentation/1.5/reference/html/#basic-mappings

Maybe the problem is, only have the method `long userId();` in the interface is not 'enough' to detect it as a proper bean → there might be on property userId() in the implementing class. But I'm just guessing…

It seems a bit unintuitive, I agree. I might take a look at the code at some point.

**Comment by thunderhook** [comment]
Records have been supported since 1.4.0, see #2014. However, they never made it into the reference guide. 

It seems that the interface interferes with records. Thanks for the detailed report, we will look into it.

**Comment by filiphr** [comment]
Everything here works as designed.

From what I can see the `TestMapper` looks like:

```java
SomeModelWithUserId fromUserIdAndSomeOtherData(UserIdProvider userIdProvider, String someOtherData);
```

Let's dissect it.

The target is `SomeModelWithUserId` which is a record, so it has two properties `userId` and `someOtherData`.
The target parameters are `UserIdProvider userIdProvider` and `String someOtherData`.
The `String someOtherData` gets mapped to the `someOtherData` of the `SomeModelWithUserId`.

The other source parameter is `UserIdProvider` which is an interface with a method `long userId()`. As MapStruct is an annotation processor we are only going to look at that interface and since the interface is not a record we are not going to use the method as a getter, since it does not follow the Java Bean specification.


Currently, there is no easy way for MapStruct to support record-style getters on interfaces. We have strict defined rules for what can be considered as a getter. It is not easy, because record-style getter is a method with no parameters returning a non void type. How do you think that we can distinquish between a getter and a regular no parameter method?

**Comment by Akatroj** [comment]
@filiphr I don't agree with this issue being closed.

>It is not easy, because record-style getter is a method with no parameters returning a non void type. How do you think that we can distinquish between a getter and a regular no parameter method?

I would assume that when the target has a `long userId` field, we could look if there's any `long userId()` method available. The method is easily distinguishable as:
* it's name is exactly the same as the target property's name and
* the method's return type matches the target property's type.

These getters could have the smallest possible priority - we look for them only if there are no other alternatives and the target field would have been left uninitialized otherwise.


I believe implementing it in this way won't break any backwards compatibility, only very rare cases when someone was _relying_ on mapstruct failing to map the field and leaving it uninitialized, which he or she shouldn't be doing anyway. If there are some other edge cases I'm not seeing, this could be introduced as an opt-in mechanism, like having to add a `@RecordInterface` annotation on the interface to be recognised by mapstruct as valid, or a `@RecordInterfaceMapper` annotation so that the mapper will look for this type of getters.


I believe records are the future of Java because of their immutability and as there are no abstract classes nor inheritance with records, using interfaces like these in the codebase is unavoidable. With how versatile MapStruct is, it was very unexpected for me that the library does not support this case. I kindly ask to reconsider your stance on this 🙂

**Comment by filiphr** [comment]
@Akatroj it isn't as easy as you think.

Let's take what you said:

>I would assume that when the target has a `long userId` field, we could look if there's any `long userId()` method available. >The method is easily distinguishable as:
>
>* it's name is exactly the same as the target property's name and
>* the method's return type matches the target property's type.

For the first part I agree with you, we can do this easily.
The second one is something that doesn't make much sense and might not be expected. e.g. why shouldn't `String userId()` be considered as well, we do provide converting from `String` to `long`.

>I believe implementing it in this way won't break any backwards compatibility, only very rare cases when someone was relying on mapstruct failing to map the field and leaving it uninitialized

Perhaps that is true, but it needs a bit of analysis. The main problem is that due to the way our code is structured implementing something like this might not be the easiest and might also lead to some performance issues in the code generation.

I also agree with you that records are good to be used. I'll reopen this to see if there is some way that we support something like this. Even if it is with an annotation on a method or a class that would treat methods like record style method

**Comment by Jouramie** [comment]
I'm in the same situation as OP. In the meantime, I ended up implementing an `AccessorNamingStrategy` where all methods without parameter a return type are considered a getter.

**Comment by rpost** [comment]
It would be nice if mapstruct supported such use case. I'm also affected in a same way as OP.

**Comment by morrica** [comment]
I don't believe this issue is limited to interfaces. I am attempting to upgrade a dependency (`com.maxmind.minfraud:minfraud:4.0.0`) and they have moved to using record style getters for all of their pojos. These are normal classes, not interfaces or records.
I wouldn't be surprised if java continues to migrate this direction in general as the `getX()/setX()` format in java has long been criticized  by some as verbose and the team behind records obviously agrees with that sentiment.

**Comment by exijn** [comment]
Hi, I'd like to explore an implementation for this issue if it is still available.

To avoid changing the default JavaBeans behavior, I'd first investigate an opt-in approach, such as an accessor-naming option or annotation that treats eligible zero-argument methods as record-style getters. I'll evaluate interactions with conversions, inherited interfaces, existing getters, and custom `AccessorNamingStrategy` implementations.

Before preparing a PR, would maintainers prefer an opt-in processor option, an annotation-based approach, or an extension through the existing accessor naming SPI?
